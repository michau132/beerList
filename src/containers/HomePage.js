import React, { Component } from 'react';
import { observer, inject, PropTypes as MobXPropTypes } from 'mobx-react';
import styled from 'styled-components';
import BeerList from '../components/BeerList';
import Header from '../components/Header';


const Main = styled.main`
  overflow-y: scroll;
  height: 100%;
`;

const StyledIsCorrectUrl = styled.p`
  color: #ff5000;
  padding: 0 40px;
`;

@inject('BeerListStore')
@observer
class HomePage extends Component {
  static propTypes ={
    BeerListStore: MobXPropTypes.observableObject.isRequired,
  }

  constructor(props) {
    super(props);
    this.scroller = React.createRef();
    this.state = {
      isCorrectUrl: true,
    };
  }


  componentDidMount() {
    const {
      BeerListStore,
      match: { params: { id } },
      location: { pathname },
    } = this.props;
    // eslint-disable-next-line no-restricted-globals
    if (!isNaN(id) || pathname === '/') {
      BeerListStore.getBeers();
    } else {
      this.setState({
        isCorrectUrl: false,
      });
    }
  }

  handleScroll = () => {
    const { BeerListStore } = this.props;
    const { loading, error, lastBeer } = BeerListStore;
    const { scroller } = this;
    if (
      scroller
      && (scroller.scrollHeight - scroller.scrollTop) < 1000
      && !loading
      && !error
      && !lastBeer
    ) {
      BeerListStore.getBeers();
    }
  }


  render() {
    const { isCorrectUrl } = this.state;
    return (
      <Main
        onScroll={this.handleScroll}
        ref={(scroller) => {
          this.scroller = scroller;
        }}
      >
        <Header />
        <BeerList />
        {!isCorrectUrl && (
        <StyledIsCorrectUrl>
          Cannot get data, provided url at the end does not match number.
          You should replace
          {' '}
          <strong>
            &quot;
            {`${window.location.hash.substring(2)}`}
          &quot;
          </strong>
            with number like
          {' '}
          <strong>&quot;12&quot;</strong>
          .
        </StyledIsCorrectUrl>
        )}
      </Main>
    );
  }
}

export default HomePage;
