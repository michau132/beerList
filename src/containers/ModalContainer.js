import React, { Component } from 'react';
import styled from 'styled-components';
import { inject, observer, PropTypes as MobXPropTypes } from 'mobx-react';
import { toJS } from 'mobx';
import CloseBtn from '../components/CloseBtn';
import Modal from '../components/Modal';

const ModalWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  overflow-x: hidden;
  overflow-y: auto;
  background-color: rgba(122,122,122, 0.5);
  padding: 10px;
  z-index: 10;
  
  @media (min-width: 1024px) {
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;

const ModalStyles = styled.div`
  background-color: #fff;
  padding: 10px;
  position: relative;
  overflow: hidden;
  ${({ loadingOrError }) => (loadingOrError ? 'height: 400px' : '')}
  
  @media (min-width: 1024px) {
    width: 910px;
  }
`;

@inject('ModalStore')
@observer
class ModalContainer extends Component {
  static propTypes = {
    ModalStore: MobXPropTypes.observableObject.isRequired,
  }

  componentDidMount() {
    const { match: { params: { id } }, ModalStore } = this.props;
    ModalStore.getBeer(id);
  }

  componentDidUpdate(prevProps) {
    const { match: { params: { id } }, ModalStore } = this.props;
    if (prevProps.match.params.id !== id) {
      ModalStore.getBeer(id);
    }
  }

  render() {
    const { ModalStore } = this.props;
    const store = toJS(ModalStore);
    const { loading, error } = store;
    return (
      <ModalWrapper>
        <ModalStyles loadingOrError={(loading || error) ? 1 : 0}>
          <CloseBtn />
          <Modal {...store} />
        </ModalStyles>
      </ModalWrapper>
    );
  }
}

export default ModalContainer;
