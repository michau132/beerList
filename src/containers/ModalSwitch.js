import React, { Component, Fragment } from 'react';
import { Switch, Route } from 'react-router-dom';
import PropTypes from 'prop-types';
import HomePage from './HomePage';
import ModalContainer from './ModalContainer';
import NotFoundPage from '../components/NotFoundPage';

class ModalSwitch extends Component {
  static propTypes = {
    location: PropTypes.shape({
      state: PropTypes.shape({
        modal: PropTypes.bool.isRequired,
      }),
    }).isRequired,
  }

  // eslint-disable-next-line react/destructuring-assignment
  previousLocation = this.props.location;


  componentDidUpdate(nextProps) {
    const { location } = this.props;
    if (
      nextProps.history.action !== 'POP'
      && (!location.state || !location.state.modal)
    ) {
      this.previousLocation = location;
    }
  }


  render() {
    const { location } = this.props;
    const path = location.pathname.substring(1);
    const isModal = !!(
      // eslint-disable-next-line no-restricted-globals
      !isNaN(path)
      || (location.state
      && location.state.modal)
    );
    return (
      <Fragment>
        <Switch location={isModal ? this.previousLocation : location}>
          <Route
            exact
            path="/"
            component={HomePage}
          />
          <Route
            path="/:id"
            exact
            component={HomePage}
          />
          <Route path="*" component={NotFoundPage} />
        </Switch>
        {isModal ? (
          <Route
            path="/:id"
            component={ModalContainer}
          />
        ) : null}

      </Fragment>
    );
  }
}

export default ModalSwitch;
