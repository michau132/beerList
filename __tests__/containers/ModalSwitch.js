import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ModalSwitch from '../../src/containers/ModalSwitch';
import HomePage from '../../src/containers/HomePage';
import ModalContainer from '../../src/containers/ModalContainer';

const props = {
  location: {
    pathname: '/2',
    state: {
      modal: true,
    },
  },
};

describe('<ModalSwitch/>', () => {
  describe('component', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<ModalSwitch {...props} />);
    });

    test('renders as expected', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('renders without crashing', () => {
      expect(wrapper).toBeDefined();
    });

    test('routes "/" to HomePage', () => {
      expect(wrapper.find('Route[exact=true][path="/"]').first().prop('component')).toBe(HomePage);
    });

    test('routes "/:id" to HomePage', () => {
      expect(wrapper.find('Route[exact=true][path="/:id"]').first().prop('component')).toBe(HomePage);
    });

    test('routes "/:id" and should be defined', () => {
      expect(wrapper.find('Route[path="/:id"]').not('[exact=true]').prop('component')).toBe(ModalContainer);
    });
  });
});
