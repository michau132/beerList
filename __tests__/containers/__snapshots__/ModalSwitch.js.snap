// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalSwitch/> component renders as expected 1`] = `
<Fragment>
  <Switch
    location={
      Object {
        "pathname": "/2",
        "state": Object {
          "modal": true,
        },
      }
    }
  >
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/:id"
    />
    <Route
      component={[Function]}
      path="*"
    />
  </Switch>
  <Route
    component={[Function]}
    path="/:id"
  />
</Fragment>
`;
